### ISSUE TITLE
Token healing (under 40 LOC)

### ISSUE THREAD

**Issue by Ayenem** [ggerganov/llama.cpp#4778]
# Feature Description

Token healing rectifies the token boundary bias in greedy tokenization. It does this by trimming and regrowing the prompt to better align with the model's tokenizer, thus enhancing generation quality. The improvement is clearest with completion models.

Debiasing token boundaries also addresses output sensitivity to prompts with trailing whitespace.

Token boundary bias is a silent performance killer that doesn't seem very well known. It has clear impact on completion quality, though I'm not sure where it would fit as a llama.cpp feature.

A more thorough explanation of the problem: [The Art of Prompt Design: Prompt Boundaries and Token Healing | by Scott Lundberg](https://towardsdatascience.com/the-art-of-prompt-design-prompt-boundaries-and-token-healing-3b2448b0be38)

# Motivation

Given a completion prompt with a partial url ending with `:`, the model might have seen the expected completion `://` as a _single_ token in training. However, the prompt's tail token `:` tells it that the next token is not `//`, and so it looks for wrong completions. Such errors compound in auto-regressive language models.

# Possible Implementation

My implementation (under 40 LOC): https://github.com/Ayenem/TokenHealer/tree/main

**Comment by cmp-nct** [comment]
That's quite interesting.

I found it remarkable that the llm's are actually able to properly generalize (so I would assume they add `//` anyway) but you can see that they struggle hard in such cases.

The problem is probably deeper than that example.
Let's take the token `e.` as in `example.`. If the prompt ends with `example` and a dot is the perfect next character then it would have to add a single `.` or `. ` resulting in a irregular token sequence.
It even touches manual biases, tokenization might provide multiple options that all lead to the same unwanted result. 


However in the example provided I find the "healing" is quite aggressive, given it changed http: into https:// - that's an actual damage to the original prompt.
Wouldn't it be possible to detect such a token-issue and handle it in live inference instead of pre-tokenization ?
As in actually testing logits after "http" and if a primary logit is "://" then use that. It's a bit like grammar sampling

**Comment by Ayenem** [comment]
@cmp-nct You make good points, your example -pun intended- is a good one. I should probably add it as a test case.

You're right about the aggressive healing, I just a merged a fix for it. I slightly bias towards preserving the original token, but that bias is still overcome when the model is really confident. So `http` doesn't get turned to `https`, but `:` is still regrown as `://`.
